scipy/sparse/_coo.py

```diff
@@ -355,7 +355,7 @@ def _coo_to_compressed(self, swap):
         nnz = len(major)
         # convert idx_dtype intc to int32 for pythran.
         # tested in scipy/optimize/tests/test__numdiff.py::test_group_columns
-        idx_dtype = self._get_index_dtype(self.coords, maxval=max(self.nnz, N))
+        idx_dtype = self._get_index_dtype(self.coords, maxval=max(self.nnz, M, N))
```

Comment:
This isn't the right fix. The issue is that our sparsetools functions currently require the shape integer scalars to have the same dtype as the index arrays. So even though int32 is the correct dtype for `indptr` and `indices` here, it can't store the `M` scalar.

We should open a separate PR to fix sparsetools.

Reply:
I removed the attempted fix for int32. We'll handle that in another PR.